src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp

```diff
@@ -90,18 +79,105 @@ class SSELoss
     gradients = (input.row(1) - input.row(0)).t();
     hessians = arma::vec(input.n_cols, arma::fill::ones);
 
-    return std::pow(ApplyL1(arma::accu(gradients)), 2) /
-        (arma::accu(hessians) + lambda);
+    // Calculates total sum of gradients and hessians.
+    gTotal = arma::accu(gradients);
+    hTotal = arma::accu(hessians);
+
+    return std::pow(ApplyL1(gTotal), 2) / (hTotal + lambda);
+  }
+
+  /**
+   * Sorts the gradients and hessians according to the sorted order of the
+   * feature.
+   *
+   * @param sortedIndices Vector of indices according to the sorted order of
+   *      the feature.
+   * @param sortedGradients Vector to store the sorted gradients.
+   * @param sortedHessians Vector to store the sorted hessians.
+   */
+  void sortGradAndHess(const arma::uvec& sortedIndices,
+                       arma::vec& sortedGradients,
+                       arma::vec& sortedHessians)
+  {
+    for (size_t i = 0; i < sortedIndices.n_elem; ++i)
+    {
+      sortedGradients[i] = gradients[sortedIndices[i]];
+      sortedHessians[i] = hessians[sortedIndices[i]];
+    }
+  }
+
+  /**
+   * Calculates the sum of hessians and gradients for the left child.
+   *
+   * @return The start index for the iteration to find the best split.
+   */
+  size_t BinaryScanInitialize(const arma::vec& sortedGradients,
+                              const arma::vec& sortedHessians)
+  {
+    size_t index = 0;
+    // Initializing the data members.
+    gLeft = 0;
+    hLeft = 0;
+    while (index < sortedHessians.n_elem - 1 &&
+           hLeft + sortedHessians[index] < minChildWeight)
+    {
+      gLeft += sortedGradients[index];
+      hLeft += sortedHessians[index++];
+    }
+    return index;
```

Comment:
I think maybe it might be better to encode the `minChildWeight` condition as part of the `XGBExactNumericSplit` class.

Reply:
For doing that, we will have to move `gLeft`, `gRight`, `hLeft`, `hRight` and `minChildWeight` to the `XGBExactNumericSplit` class. We can't move `minChildWeight` there because there is no way to store it as we are not instantiating `NumericSplit` classes.

I know it makes more sense to have the condition within the loop itself, but I couldn't think of a cleaner way of doing it. Do you have something in your mind?

Reply:
Ahh, I overlooked that.  But, note that the API you are using here is different than what we came up with before for `BinaryScanInitialize()` and `BinaryStep()` (now they return `size_t` and `bool`, respectively).  I think that change is fine, but can you also update the other loss functions and `BestBinaryNumericSplit` accordingly?